### PR TITLE
Add warnings if loading translations too early

### PR DESCRIPTION
**Pull request by swissspidy**
In my quick local testing I already identified a few plugins doing `load_plugin_textdomain()` wrong:

* [ ] [RTL Tester](https://wordpress.org/plugins/rtl-tester/)
	* Not really maintained anymore anyway
* [x] [Two Factor](https://wordpress.org/plugins/two-factor/)
	* https://github.com/WordPress/two-factor/pull/608
* [x] Elementor
	* https://github.com/elementor/elementor/pull/27200 
* [ ] Litespeed Cache
	* https://github.com/litespeedtech/lscache_wp/pull/738

Plugins doing just-in-time loading wrong:

* [x] WooCommerce
	* https://github.com/woocommerce/woocommerce/pull/47113
* [ ] BuddyPress
	* Their theme integration also causes Twenty Twenty-Four translations to be loaded too early

---

**To-do:**

*  Improve wording for the `_load_textdomain_just_in_time` case, mentioning that it's typically caused by unintended side effects in a plugin or theme.
* **What about themes?** They usually do stuff in `after_setup_theme`, which runs right before `init` and current user setup. Probably all classic themes will be affected by this.

---

Strangely, when one of these plugins and Query Monitor is active, there is some sort of infinite loop happening and the site crashes with a 500. @johnbillion Any thoughts why QM might run in circles here?

---

Trac ticket: https://core.trac.wordpress.org/ticket/44937

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by johnbillion** [comment]
I'll take a look at what's happening in QM. I would recommend testing this out with some multilingual plugins if you haven't already. In particular WPML (and add-on plugins for it) has been a bit of a pain in the past with regard to loading translations too early.

**Comment by johnbillion** [comment]
I'm not seeing an infinite loop in my local testing with RTL Tester. I am concerned about calling `__()` from within the load textdomain functions though. Do we do that anywhere else?

**Comment by swissspidy** [comment]
OK here's how to reproduce the error. Even if you can't reproduce it so far, it seems logical:

1. Set user profile to something else than en_US
2. Activate RTL Tester
3. Activate Query Monitor
4. RTL Tester triggers the `_doing_it_wrong` in `load_plugin_textdomain()`
5. This calls `QM_Collector_Doing_It_Wrong::action_doing_it_wrong_run()` which uses `__()`
6. This in turn triggers the `_doing_it_wrong` in `_load_textdomain_just_in_time()`
7. This calls `QM_Collector_Doing_It_Wrong::action_doing_it_wrong_run()` again...

Adding some caching to only warn about a specific text domain once would solve this. Like a `_doing_it_wrong_once` function or so.

> I am concerned about calling `__()` from within the load textdomain functions though. Do we do that anywhere else?

I don't think so. In theory it shouldn't cause any issues as the `__()` call and the `load_*_textdomain()` call use different domains. And the QM issue is only because it taps into the `_doing_it_wrong`.

**Comment by swissspidy** [comment]
Actually, it's more like QM would need to somehow unhook `QM_Collector_Doing_It_Wrong` in itself because it is triggering `_doing_it_wrong` itself.

**Comment by johnbillion** [comment]
Good spot, thanks. Temporarily unhooking itself is probably best otherwise there's no way to know what else might trigger a `_doing_it_wrong` call within any of the functions that are called from a doing it wrong handler.

**Comment by swissspidy** [comment]
Looking into warnings emitted by themes, such as `twentytwentyone`.

Some observations:

* Themes usually trigger just-in-time loading on `after_setup_theme`
* In the admin, the current user is set up right after `setup_theme` because of the `load_default_textdomain()` call in `wp-settings.php`
* Checking for `after_setup_theme` to emit a warning might be a better solution

> I am concerned about calling `__()` from within the load textdomain functions though.

At this point `determine_locale()` will already have been called, and the current user thus setup. So I don't see any other side effect happening because of this.

File: src/wp-includes/l10n.php [hunk 1]
```diff
@@ -999,6 +999,19 @@ function load_plugin_textdomain( $domain, $deprecated = false, $plugin_rel_path
...
 	}
 
+	if ( ! doing_action( 'after_setup_theme' ) && ! did_action( 'after_setup_theme' ) ) {
+		_doing_it_wrong(
+			__FUNCTION__,
+			sprintf(
+				/* translators: 1: The text domain. 2: 'init'. */
+				__( 'Attempted to load translations for the %1$s domain too early. Translations should be loaded after the %2$s action has fired, to ensure that the current user has been already set up.' ),
+				'<code>' . $domain . '</code>',
+				'<code>init</code>'
```
Comment (swissspidy) [inline src/wp-includes/l10n.php]:
```suggestion
				'<code>after_setup_theme</code>'
```

File: src/wp-includes/l10n.php [hunk 2]
```diff
@@ -1051,6 +1064,19 @@ function load_muplugin_textdomain( $domain, $mu_plugin_rel_path = '' ) {
...
```
Comment (swissspidy) [inline src/wp-includes/l10n.php]:
```suggestion
				'<code>after_setup_theme</code>'
```

File: src/wp-includes/l10n.php [hunk 3]
```diff
@@ -1094,6 +1120,19 @@ function load_theme_textdomain( $domain, $path = false ) {
...
```
Comment (swissspidy) [inline src/wp-includes/l10n.php]:
```suggestion
				'<code>after_setup_theme</code>'
```

File: src/wp-includes/l10n.php [hunk 4]
```diff
@@ -1373,6 +1412,20 @@ function _load_textdomain_just_in_time( $domain ) {
...
 	}
+
+	if ( ! doing_action( 'after_setup_theme' ) && ! did_action( 'after_setup_theme' ) ) {
+		_doing_it_wrong(
+			__FUNCTION__,
+			sprintf(
+				/* translators: 1: The text domain. 2: 'init'. */
+				__( 'Translation loading for the %1$s domain was triggered too early. This is usually an indicator for some code running too early.' ),
+				'<code>' . $domain . '</code>',
+				'<code>init</code>'
```
Comment (swissspidy) [inline src/wp-includes/l10n.php]:
```suggestion
				'<code>after_setup_theme</code>'
```

File: src/wp-includes/l10n.php [hunk 5]
```diff
@@ -1373,6 +1412,20 @@ function _load_textdomain_just_in_time( $domain ) {
...
+			),
+			'6.6.0'
```
Comment (swissspidy) [inline src/wp-includes/l10n.php]:
```suggestion
			'6.7.0'
```

File: src/wp-includes/l10n.php [hunk 6]
```diff
@@ -1094,6 +1120,19 @@ function load_theme_textdomain( $domain, $path = false ) {
...
+	if ( ! doing_action( 'after_setup_theme' ) && ! did_action( 'after_setup_theme' ) ) {
+		_doing_it_wrong(
+			__FUNCTION__,
+			sprintf(
+				/* translators: 1: The text domain. 2: 'init'. */
+				__( 'Attempted to load translations for the %1$s domain too early. Translations should be loaded after the %2$s action has fired, to ensure that the current user has been already set up.' ),
+				'<code>' . $domain . '</code>',
+				'<code>init</code>'
+			),
+			'6.6.0'
```
Comment (swissspidy) [inline src/wp-includes/l10n.php]:
```suggestion
			'6.7.0'
```

File: src/wp-includes/l10n.php [hunk 7]
```diff
@@ -999,6 +999,19 @@ function load_plugin_textdomain( $domain, $deprecated = false, $plugin_rel_path
...
```
Comment (swissspidy) [inline src/wp-includes/l10n.php]:
```suggestion
			'6.7.0'
```

File: src/wp-includes/l10n.php [hunk 8]
```diff
@@ -1051,6 +1064,19 @@ function load_muplugin_textdomain( $domain, $mu_plugin_rel_path = '' ) {
...
```
Comment (swissspidy) [inline src/wp-includes/l10n.php]:
```suggestion
			'6.7.0'
```

File: src/wp-includes/l10n.php [hunk 9]
```diff
@@ -1373,6 +1412,20 @@ function _load_textdomain_just_in_time( $domain ) {
 	if ( ! $path ) {
 		return false;
 	}
+
+	if ( ! doing_action( 'after_setup_theme' ) && ! did_action( 'after_setup_theme' ) ) {
+		_doing_it_wrong(
+			__FUNCTION__,
+			sprintf(
+				/* translators: 1: The text domain. 2: 'init'. */
+				__( 'Translation loading for the %1$s domain was triggered too early. This is usually an indicator for some code running too early.' ),
```
Comment (swissspidy) [inline src/wp-includes/l10n.php]:
```suggestion
				__( 'Translation loading for the %1$s domain was triggered too early. This is usually an indicator for some code in the plugin/theme running too early.' ),
```

File: src/wp-includes/l10n.php [hunk 10]
```diff
@@ -999,6 +999,19 @@ function load_plugin_textdomain( $domain, $deprecated = false, $plugin_rel_path
 		return false;
 	}
 
+	if ( ! doing_action( 'after_setup_theme' ) && ! did_action( 'after_setup_theme' ) ) {
+		_doing_it_wrong(
+			__FUNCTION__,
+			sprintf(
+				/* translators: 1: The text domain. 2: 'init'. */
```
Comment (swissspidy) [inline src/wp-includes/l10n.php]:
```suggestion
				/* translators: 1: The text domain. 2: 'after_setup_theme'. */
```

File: src/wp-includes/l10n.php [hunk 11]
```diff
@@ -1051,6 +1064,19 @@ function load_muplugin_textdomain( $domain, $mu_plugin_rel_path = '' ) {
...
```
Comment (swissspidy) [inline src/wp-includes/l10n.php]:
```suggestion
				/* translators: 1: The text domain. 2: 'after_setup_theme'. */
```

File: src/wp-includes/l10n.php [hunk 12]
```diff
@@ -1094,6 +1120,19 @@ function load_theme_textdomain( $domain, $path = false ) {
...
```
Comment (swissspidy) [inline src/wp-includes/l10n.php]:
```suggestion
				/* translators: 1: The text domain. 2: 'after_setup_theme'. */
```

File: src/wp-includes/l10n.php [hunk 13]
```diff
@@ -1373,6 +1412,20 @@ function _load_textdomain_just_in_time( $domain ) {
 	if ( ! $path ) {
 		return false;
 	}
+
+	if ( ! doing_action( 'after_setup_theme' ) && ! did_action( 'after_setup_theme' ) ) {
+		_doing_it_wrong(
+			__FUNCTION__,
+			sprintf(
+				/* translators: 1: The text domain. 2: 'init'. */
```
Comment (swissspidy) [inline src/wp-includes/l10n.php]:
```suggestion
				/* translators: 1: The text domain. 2: 'after_setup_theme'. */
```

File: src/wp-includes/l10n.php [hunk 14]
```diff
@@ -999,6 +999,19 @@ function load_plugin_textdomain( $domain, $deprecated = false, $plugin_rel_path
 		return false;
 	}
 
+	if ( ! doing_action( 'after_setup_theme' ) && ! did_action( 'after_setup_theme' ) ) {
+		_doing_it_wrong(
+			__FUNCTION__,
+			sprintf(
+				/* translators: 1: The text domain. 2: 'after_setup_theme'. */
+				__( 'Attempted to load translations for the %1$s domain too early. Translations should be loaded after the %2$s action has fired, to ensure that the current user has been already set up.' ),
```
Comment (swissspidy) [inline src/wp-includes/l10n.php]:
```suggestion
				__( 'Attempted to load translations for the %1$s domain too early. Translations should be loaded after the %2$s action has fired, to ensure that the current user is already set up.' ),
```

File: src/wp-includes/l10n.php [hunk 15]
```diff
@@ -1051,6 +1064,19 @@ function load_muplugin_textdomain( $domain, $mu_plugin_rel_path = '' ) {
...
```
Comment (swissspidy) [inline src/wp-includes/l10n.php]:
```suggestion
				__( 'Attempted to load translations for the %1$s domain too early. Translations should be loaded after the %2$s action has fired, to ensure that the current user is already set up.' ),
```

File: src/wp-includes/l10n.php [hunk 16]
```diff
@@ -1094,6 +1120,19 @@ function load_theme_textdomain( $domain, $path = false ) {
...
```
Comment (swissspidy) [inline src/wp-includes/l10n.php]:
```suggestion
				__( 'Attempted to load translations for the %1$s domain too early. Translations should be loaded after the %2$s action has fired, to ensure that the current user is already set up.' ),
```

File: src/wp-includes/l10n.php [hunk 17]
```diff
@@ -1373,6 +1412,20 @@ function _load_textdomain_just_in_time( $domain ) {
...
 	}
+
+	if ( ! doing_action( 'after_setup_theme' ) && ! did_action( 'after_setup_theme' ) ) {
+		_doing_it_wrong(
+			__FUNCTION__,
+			sprintf(
+				/* translators: 1: The text domain. 2: 'after_setup_theme'. */
+				__( 'Translation loading for the %1$s domain was triggered too early. This is usually an indicator for some code in the plugin/theme running too early.' ),
+				'<code>' . $domain . '</code>',
+				'<code>after_setup_theme</code>'
```
Comment (swissspidy) [inline src/wp-includes/l10n.php]:
```suggestion
				/* translators: %s: The text domain. */
				__( 'Translation loading for the %s domain was triggered too early. This is usually an indicator for some code in the plugin or theme running too early.' ),
				'<code>' . $domain . '</code>'
```

**Comment by swissspidy** [comment]
@ocean90 @SergeyBiryukov @johnbillion Is this something one of you could provide feedback on? I think it's a nice safeguard that could still land before the beta, but it also might be a bit noisy.

**Review by johnbillion — APPROVED** [review]

**Comment by johnbillion** [comment]
This could be noisy but it might be interesting to see if any feedback about that is raised during beta.

**Comment by swissspidy** [comment]
Committed in https://core.trac.wordpress.org/changeset/59127

**Pull request closed without merging**